source/extensions/filters/http/cache/cache_insert_queue.cc

```diff
@@ -0,0 +1,188 @@
+#include "source/extensions/filters/http/cache/cache_insert_queue.h"
+
+#include "source/common/buffer/buffer_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+
+class CacheInsertChunk {
+public:
+  // Sends a chunk to the cache.
+  // on_complete is called when the cache completes the operation.
+  virtual void send(InsertContext& context,
+                    std::function<void(bool ready, bool end_stream, size_t sz)> on_complete) PURE;
+
+  virtual ~CacheInsertChunk() = default;
+};
+
+class CacheInsertChunkBody : public CacheInsertChunk {
+public:
+  CacheInsertChunkBody(const Buffer::Instance& buffer, bool end_stream)
+      : buffer_(buffer), end_stream_(end_stream) {}
+
+  void send(InsertContext& context,
+            std::function<void(bool ready, bool end_stream, size_t sz)> on_complete) override {
+    size_t sz = buffer_.length();
+    context.insertBody(
+        std::move(buffer_),
+        [on_complete, end_stream = end_stream_, sz](bool ready) {
+          on_complete(ready, end_stream, sz);
+        },
+        end_stream_);
+  }
+
+private:
+  Buffer::OwnedImpl buffer_;
+  const bool end_stream_;
+};
+
+class CacheInsertChunkTrailers : public CacheInsertChunk {
+public:
+  explicit CacheInsertChunkTrailers(const Http::ResponseTrailerMap& trailers)
+      : trailers_(Http::ResponseTrailerMapImpl::create()) {
+    Http::ResponseTrailerMapImpl::copyFrom(*trailers_, trailers);
+  }
+
+  void send(InsertContext& context,
+            std::function<void(bool ready, bool end_stream, size_t sz)> on_complete) override {
+    // While zero isn't technically true for the size of trailers, it doesn't
+    // matter at this point because watermarks after the stream is complete
+    // aren't useful.
+    context.insertTrailers(*trailers_, [on_complete](bool ready) { on_complete(ready, true, 0); });
+  }
+
+private:
+  std::unique_ptr<Http::ResponseTrailerMap> trailers_;
+};
+
+// We capture the values from encoder_callbacks, and the callbacks as lambdas, so
+// that we don't have to keep checking whether the filter has been deleted.
+// When the filter *is* deleted, we simply replace all the callback lambdas with no-op.
```

Comment:
wouldn't it be easier to hold encoder_callbacks as an `OptRef<Http::StreamEncoderFilterCallbacks>` and reset that when the filter is deleted, and not bother to hold the 4 things computed from encoder_callbacks?